### ISSUE TITLE
Renewing Certificates requires certificate url

### ISSUE THREAD

**Issue by cristiangraz** [ericchiang/letsencrypt#11]
Do you have any examples for renewing certificates? From what I can tell, you need the certificate URL returned when requesting the certificate. The library doesn't currently return that. 

I'd be helpful to have the Location and Content-Location headers returned from calling NewCertificate. I'd like to store those. What do you think about changing the method signature to:

```
NewCertificate(accountKey interface{}, csr *x509.CertificateRequest) (*x509.Certificate, Location string, ContentLocation string, error) {
```

or

```
NewCertificate(accountKey interface{}, csr *x509.CertificateRequest) (*x509.Certificate, http.Header, error) {
```

I'm having trouble getting boot2docker to actually forward requests to my host machine so I can verify challenges and get to that step, so haven't been able to verify (otherwise I'd submit a PR).

**Comment by ericchiang** [comment]
It looks like the URL to get the certificate is in the "Location" header [according to the spec](https://tools.ietf.org/html/draft-ietf-acme-acme-01#section-6.6). I think this is the same as the challenge, and you should be able to get it from the challenge ID.

Last I checked, boulder didn't implement getting the challenges associated with the user as laid out in the spec [here](https://tools.ietf.org/html/draft-ietf-acme-acme-01#section-5.3). Let me look that over again and see if there's a way to list all challenges a user has open, and in turn reissue certificates.

**Comment by cristiangraz** [comment]
Thanks, Eric. From what I understand of section 6.6 (page 32), `Content-Location` is the URL to the current certificate. Any future requests to that URL will always return the exact same certificate. `Location` is the URL to hit if you want a new certificate (and how you would be able to handle renewals).

I also found a potential bug. I'm going to list here as it's relevant to the conversation. If you look in section 6.6 of the spec on page 31:

```
 If the CA decides to issue a certificate, then the server creates a
   new certificate resource and returns a URI for it in the Location
   header field of a 201 (Created) response.

   HTTP/1.1 201 Created
   Location: https://example.com/acme/cert/asdf

   If the certificate is available at the time of the response, it is
   provided in the body of the response.  If the CA has not yet issued
   the certificate, the body of this response will be empty.  The client
   should then send a GET request to the certificate URI to poll for the
   certificate.  As long as the certificate is unavailable, the server
   MUST provide a 202 (Accepted) response and include a Retry-After
   header to indicate when the server believes the certificate will be
   issued (as in the example above).
```

Right now, NewCertificate always assumes a certificate is immediately available in the response body, when there's a chance a Retry-After header will be sent and some polling may be required to actually obtain the certificate.

That's why I was thinking it may be beneficial for NewCertificate to return the URLs or headers. If NewCertificate isn't able to return a certificate immediately, but no error was thrown, the user can handle polling the URL until the certificate is ready. Or we can even add in a convenience function in the library. In this case the headers may be helpful since the user would need Location, Content-Location, and Retry-After headers after the call to NewCertificate.

Lego handles this by sleeping for the Retry-After period and trying again once. Since the Retry-After period could be long, I don't particularly like this approach since users using this library are piecing together their own functionality. Here is the relevant code however: https://github.com/xenolf/lego/blob/master/acme/client.go#L575-L614

Still trying to get boulder to verify challenges on my mac, but if I can get that working I'm happy to work on this and submit a PR once we come up with how this should work.

**Comment by ericchiang** [comment]
Thanks for the write-up!

Yeah there seem to be two issues
1. If the cert isn't immediately issues this client will return an error. This needs to be rectified and is probably best for another issue.
2. How to reissue a certificate after the initial creation.

For point 2, I understand that the create certificate request returns a URI. However, you should be able to get the list certificates from your authorization object. Ideally, you wouldn't need to store the URL, you'd just need to have your private key. The steps would be

```
privKey -> registration object  -> certificate
```

I don't know if bolder has implemented this part of the spec. This needs to be checked.

The certificates field on the registration object I'm talking about is discussed [here](https://tools.ietf.org/html/draft-ietf-acme-acme-01#section-5.2).

```
    certificates (optional, string):  A URI from which a list of
      certificates issued for this account can be fetched via a GET
      request.  The result of the GET request MUST be a JSON object
      whose "certificates" field is an array of strings, where each
      string is the URI of a certificate.  The server SHOULD NOT include
      expired certificates.

   {
     "resource": "new-reg",
     "contact": [
       "mailto:cert-admin@example.com",
       "tel:+12025551212"
     ],
     "agreement": "https://example.com/acme/terms",
     "authorizations": "https://example.com/acme/reg/1/authz",
     "certificates": "https://example.com/acme/reg/1/cert",
   }
```

**Comment by cristiangraz** [comment]
Thanks, Eric. I did see that (I will test it out shortly). However, we're already managing thousands of domains and growing. Since the renewal timeline is so short, having to go through that process each time is going to add unnecessary overhead and really slow things down when we need to process hundreds of renewals per day.

Would you be opposed to returning the URL(s) for the certificate? I also realize #1 is a separate issue, but this would help solve that one as well. I'm happy to do the work to get a PR submitted.

**Comment by ericchiang** [comment]
Cool. Thanks for clarifying your use case.

I'm fine to make an API change to return the URL as well. Maybe `NewCertificate` should return a `CertificateResponse` struct with other data (possibly added in the future) as well?

I'd like to make the function match up with [`NewAuthorization`](https://godoc.org/github.com/ericchiang/letsencrypt#Client.NewAuthorization) since they both return URLs. So whatever the change made, these two should look similar.

Did you have anything in mind?

**Comment by cristiangraz** [comment]
Check out this branch:
https://github.com/blacklightcms/letsencrypt/tree/feature/certificate-refactor

If you like the direction, I'll create the tests and we can move the rest of the conversation to the PR.

Details:
- NewCertificate now returns CertificateResponse struct
- NewCertificate now supports asynchronous certificates (Not yet in Boulder, but based on spec)
- If NewCertificate call results in a synchronous certificate, CertificateResponse.Certificate is populated with *x509.Certificate. URI and StableURI are both set. The issuer url is set based on the rel="up" Link header. (see below for more details
- If NewCertificate call results in an empty response body, CertificateResponse is populated with RetryAfter and URI.
- This sets up an easy implementation for certificate renewals (to be implemented later)

Methods on Certificate Response:
- CertificateResponse contains two methods: Retry and Bundle
- Retry will handle implementation details to retry the call to CertificateResponse.URI. User triggers it based on number of seconds passed in CertificateResponse.RetryAfter. Behavior here is the same as NewCertificate depending on if the certificate is available or requires another try.
- Bundle allows for bundling the certificate with the cross-signed Issuer. This is a convenience method and puts the burden on the user to call, but the implementation details are covered in the library.

Resources:
- https://github.com/letsencrypt/boulder/issues/934
- https://letsencrypt.org/certificates/

Thanks for all your prompt feedback!

**Comment by cristiangraz** [comment]
I just added RenewCertificate functionality to the branch. Both NewCertificate and RenewCertificate now mostly do the same since they overlap for most of their functionality. They also both support asynchronous certificates when boulder gets around to adding that.

I've been testing the workflow of this branch for the last day and so far it seems pretty good. Let me know if you have any feedback, otherwise I'm going to get started on unit tests so I can get a PR in for you to review.

**Comment by ericchiang** [comment]
You're free to make a pull request even if it lacks unit test. That way we can track everything in one repo :)

**Comment by cristiangraz** [comment]
Working on tests. Will submit a pull request within the next day.

**Comment by ericchiang** [comment]
@cristiangraz if you make a pull request, I can review the changes before you write the tests.

**Comment by cristiangraz** [comment]
@ericchiang sorry I misunderstood. Pull request is available here: https://github.com/ericchiang/letsencrypt/pull/13

**Comment by cristiangraz** [comment]
Fixed in #14.
